### PR TITLE
fivetran-destination: Support reporting `primary_key` from a `DescribeTable` request

### PR DESCRIPTION
**Pull request by ParkMyCar**
When we get a `CreateTableRequest` from Fivetran they include a list of primary keys. Later when they run a `DescribeTableRequest` they expect this information to roundtrip, i.e. if they made column "foo" a primary key `DescribeTable` should indicate it as one. If `DescribeTable` does not match their expectations they run an `ALTER TABLE ...` request to update the table.

Materialize does not support primary keys, so to prevent Fivetran from running an `ALTER TABLE ...` request (which we also don't yet support), we store a magic string as a SQL COMMENT on the column to indicate whether or not it's a primary key.

This is a HACK and is not something we want to ship with, instead we should support `PRIMARY KEY ... NOT ENFORCED` and include whether or not something is a primary key in the `mz_columns` table or something similar. But in the interest of unblocking testing, this is the fastest thing we could come up with.

### Motivation

Reported as a blocker by Fivetran: https://materializeinc.slack.com/archives/C060KAR4802/p1707249394772389?thread_ts=1707202803.646169&cid=C060KAR4802

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: src/fivetran-destination/src/destination/ddl.rs [hunk 1]
```diff
@@ -49,29 +53,35 @@ pub async fn handle_describe_table(
     };
 
     let columns = {
+        let stmt = format!(
+            r#"SELECT
+                   name,
+                   type_oid,
+                   type_mod,
+                   CASE WHEN coms.comment = {magic_comment} THEN True ELSE false END AS primary_key
```
Comment (mjibson) [inline src/fivetran-destination/src/destination/ddl.rs:61]:
```suggestion
                   coms.comment = $1 AS primary_key
```

File: src/fivetran-destination/src/destination/ddl.rs [hunk 2]
```diff
@@ -49,29 +53,35 @@ pub async fn handle_describe_table(
...
+               FROM mz_columns AS cols
+               JOIN mz_internal.mz_comments AS coms
+               ON cols.id = coms.id AND cols.position = coms.object_sub_id
+               WHERE cols.id = 'u1'"#,
```
Comment (mjibson) [inline src/fivetran-destination/src/destination/ddl.rs:65]:
```suggestion
               WHERE cols.id = $2"#;
```

File: src/fivetran-destination/src/destination/ddl.rs [hunk 3]
```diff
@@ -49,29 +53,35 @@ pub async fn handle_describe_table(
...
+
         let rows = client
-            .query(
-                r#"SELECT name, type_oid, type_mod
-                   FROM mz_columns c
-                   WHERE c.id = $1
-                "#,
-                &[&table_id],
-            )
+            .query(&stmt, &[&table_id])
```
Comment (mjibson) [inline src/fivetran-destination/src/destination/ddl.rs:70]:
```suggestion
            .query(&stmt, &[PRIMARY_KEY_MAGIC_STRING, &table_id])
```

File: src/fivetran-destination/src/destination/ddl.rs [hunk 4]
```diff
@@ -49,29 +53,35 @@ pub async fn handle_describe_table(
     };
 
     let columns = {
+        let stmt = format!(
+            r#"SELECT
```
Comment (mjibson) [inline src/fivetran-destination/src/destination/ddl.rs:57]:
```suggestion
        let stmt = r#"SELECT
```

File: src/fivetran-destination/src/destination/ddl.rs [hunk 5]
```diff
@@ -49,29 +53,35 @@ pub async fn handle_describe_table(
...
+                   name,
+                   type_oid,
+                   type_mod,
+                   CASE WHEN coms.comment = {magic_comment} THEN True ELSE false END AS primary_key
+               FROM mz_columns AS cols
+               JOIN mz_internal.mz_comments AS coms
+               ON cols.id = coms.id AND cols.position = coms.object_sub_id
+               WHERE cols.id = 'u1'"#,
+            magic_comment = escape::escape_literal(PRIMARY_KEY_MAGIC_STRING),
+        );
```
Comment (mjibson) [inline src/fivetran-destination/src/destination/ddl.rs:67]:
```suggestion
```

File: src/fivetran-destination/src/destination/ddl.rs [hunk 6]
```diff
@@ -109,23 +121,31 @@ pub async fn handle_create_table(request: CreateTableRequest) -> Result<(), OpEr
...
     client.batch_execute(&sql).await?;
 
+    // TODO(parkmycar): This is an ugly hack!
+    //
+    // If Fivetran creates a table with primary keys, it expects a DescribeTableRequest to report
+    // those columns as primary keys. But Materialize doesn't support primary keys, so we need to
+    // store this metadata somewhere else. For now we do it in a COMMENT.
+    for column_name in primary_key_columns {
+        let stmt = format!(
+            "COMMENT ON COLUMN {qualified_table_name}.{column_name} IS {magic_comment}",
```
Comment (mjibson) [inline src/fivetran-destination/src/destination/ddl.rs:140]: Also make this a param.

**Review by mjibson — COMMENTED** [review]
Can you make these params instead of format strings?

File: src/fivetran-destination/src/destination/ddl.rs (hunk 6 above)
Comment (ParkMyCar) [reply src/fivetran-destination/src/destination/ddl.rs:137]: `COMMENT ON` doesn't support specifying the comment with a param, it needs to be a string literal. This is also the case in Postgres, e.g. the following fails:
```
let mut client = postgres::Client::connect("host=localhost user=parker.timmerman", NoTls)?;

client.execute("CREATE TABLE t1 (a int)", &[])?;
client.execute("COMMENT ON COLUMN t1.a IS $1", &[&"hello world"])?;
```

File: src/fivetran-destination/src/destination/ddl.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/fivetran-destination/src/destination/ddl.rs]: Done!

File: src/fivetran-destination/src/destination/ddl.rs (hunk 3 above)
Comment (ParkMyCar) [reply src/fivetran-destination/src/destination/ddl.rs]: Done!

File: src/fivetran-destination/src/destination/ddl.rs (hunk 2 above)
Comment (ParkMyCar) [reply src/fivetran-destination/src/destination/ddl.rs]: Whoops, this was left over from testing, fixed!

File: src/fivetran-destination/src/destination/ddl.rs [hunk 7]
```diff
@@ -82,6 +89,8 @@ pub async fn handle_describe_table(
         columns
     };
 
+    tracing::info!(?columns, "our response");
```
Comment (mjibson) [inline src/fivetran-destination/src/destination/ddl.rs:92]: Is this leftover debug?

**Review by mjibson — APPROVED** [review]

File: src/fivetran-destination/src/destination/ddl.rs (hunk 7 above)
Comment (ParkMyCar) [reply src/fivetran-destination/src/destination/ddl.rs:92]: Good catch, thanks!

**Pull request merged**